### PR TITLE
Add initial cluster data version support

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
Need to integrate cluster data version storage with IM
#### Change overview
Update ReadHandler/Engine to retrieve version from ember, and process version in read client side.

#### Testing
Manually observe the data version change using IM read/write interaction.

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]

**Review by mrjerryjohns — CHANGES_REQUESTED** [review]
So given our discussions, the approach in this PR will need to change substantially to pivot away from needing to rely on persistence for storing data versions, to something more ephemeral.

File: examples/tv-app/linux/AppImpl.cpp
Comment (bzbarsky-apple) [inline examples/tv-app/linux/AppImpl.cpp:150]: Where are these macros defined?

File: src/app/BufferedReadCallback.cpp
Comment (bzbarsky-apple) [inline src/app/BufferedReadCallback.cpp:259]: I don't understand how that can possibly be safe.  We're storing a pointer to ... where?  How long will the memory it points to live?

File: src/app/ConcreteClusterPath.h
Comment (bzbarsky-apple) [inline src/app/ConcreteClusterPath.h:25]: As a followup: we should get rid of the `ClusterPath` struct in binding manager and use this instead, right?

File: src/app/ReadClient.cpp
Comment (bzbarsky-apple) [inline src/app/ReadClient.cpp:532]: Yeah, see, this is passing a pointer into the stack of this function.  That is NOT a pointer we want callees saving.

File: src/app/ReadClient.h
Comment (bzbarsky-apple) [inline src/app/ReadClient.h:116]: Which means what?  If it's null, what does that mean?  Needs to be clearly documented.

File: src/app/ReadClient.h
Comment (bzbarsky-apple) [inline src/app/ReadClient.h:122]: At the very least, that should be a `const DataVersion *`, becase we don't want callees mutating it.

But also, maybe this should actually be `const Optional<DataVersion> &` instead of a pointer?  Might prevent the problems we have above with consumers storing it....

File: src/app/reporting/Engine.cpp
Comment (bzbarsky-apple) [inline src/app/reporting/Engine.cpp:27]: Where is that used?  I don't see it being used.

File: src/app/util/attribute-storage.cpp
Comment (bzbarsky-apple) [inline src/app/util/attribute-storage.cpp:98]: Shouldn't this be an array of `DataVersion`?

File: src/app/util/af-types.h
Comment (bzbarsky-apple) [inline src/app/util/af-types.h:120]:
```suggestion
    chip::DataVersion * const version;
```

File: src/app/util/ember-compatibility-functions.cpp
Comment (bzbarsky-apple) [inline src/app/util/ember-compatibility-functions.cpp:251]: Is there a reason to not pass a `const ConcreteClusterPath &` here?

File: src/app/util/ember-compatibility-functions.cpp
Comment (bzbarsky-apple) [inline src/app/util/ember-compatibility-functions.cpp:258]: Please add an accessor on EmberAfCluster() instead of making assumptions about how to get from one to the value.

File: src/app/util/ember-compatibility-functions.cpp
Comment (bzbarsky-apple) [inline src/app/util/ember-compatibility-functions.cpp:262]: Again, pass `ConcreteClusterPath`?

File: src/app/util/ember-compatibility-functions.cpp
Comment (bzbarsky-apple) [inline src/app/util/ember-compatibility-functions.cpp:272]: Again, this should be done via a method on EmberAfCluster.  We're not implementing in C; let's not be limited to the things C could do here.

File: src/app/util/ember-compatibility-functions.cpp
Comment (bzbarsky-apple) [inline src/app/util/ember-compatibility-functions.cpp:267]:
```suggestion
        ChipLogError(DataManagement, "Path: Endpoint %" PRIx16 ", Cluster " ChipLogFormatMEI " not found in IncreaseClusterDataVersion!",
```

File: src/app/zap-templates/templates/app/endpoint_config.zapt
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/endpoint_config.zapt:40]: Why is that cast needed?  I don't think it's needed...

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/ConcreteClusterPath.h
Comment (yunhanw-google) [reply src/app/ConcreteClusterPath.h]: this file is accidental check-in to this pr, just remove it, thanks

File: examples/tv-app/linux/AppImpl.cpp
Comment (yunhanw-google) [reply examples/tv-app/linux/AppImpl.cpp]: not defined, removed, thanks

File: src/app/reporting/Engine.cpp
Comment (yunhanw-google) [reply src/app/reporting/Engine.cpp:27]: not used, removed.

File: src/app/util/attribute-storage.cpp
Comment (yunhanw-google) [reply src/app/util/attribute-storage.cpp:98]: done, thanks

File: src/app/util/af-types.h
Comment (yunhanw-google) [reply src/app/util/af-types.h:120]: done, thanks

File: src/app/util/attribute-storage.cpp
Comment (yunhanw-google) [reply src/app/util/attribute-storage.cpp:98]: for (int cluster = 0; cluster < GENERATED_CLUSTER_COUNT; cluster++)
    {
        Crypto::DRBG_get_bytes(reinterpret_cast<uint8_t *>(clusterVersions[cluster]), 4);
    }
random values are set here in line 136

File: src/app/ReadClient.h
Comment (yunhanw-google) [reply src/app/ReadClient.h:116]: updated, thanks, it would be nullptr when the value of EnableTagCompression in the Path field is true.

File: src/app/ReadClient.h
Comment (yunhanw-google) [reply src/app/ReadClient.h:122]: Updated with const DataVersion *, would decide if to switch const Optional<DataVersion> & in follow-up, thanks

File: src/app/util/ember-compatibility-functions.cpp
Comment (yunhanw-google) [reply src/app/util/ember-compatibility-functions.cpp:261]: have removed ConcreteClusterPath, it seems there is no need to use that one.

File: src/app/util/ember-compatibility-functions.cpp
Comment (yunhanw-google) [reply src/app/util/ember-compatibility-functions.cpp:250]: it seems there is no need to use ConcreteClusterPath. I have removed ConcreteClusterPath file.

File: src/app/util/ember-compatibility-functions.cpp
Comment (yunhanw-google) [reply src/app/util/ember-compatibility-functions.cpp:257]: Done, thanks

File: src/app/util/ember-compatibility-functions.cpp
Comment (yunhanw-google) [reply src/app/util/ember-compatibility-functions.cpp:266]: Done, thanks

File: src/app/util/ember-compatibility-functions.cpp
Comment (yunhanw-google) [reply src/app/util/ember-compatibility-functions.cpp:267]: Done, thanks

File: src/app/util/ember-compatibility-functions.cpp
Comment (yunhanw-google) [reply src/app/util/ember-compatibility-functions.cpp:271]: EmberAfCluster is typedef struct...we cannot add member function inside EmberAfCluster

File: src/app/ReadClient.cpp
Comment (yunhanw-google) [reply src/app/ReadClient.cpp:532]: Just update with Optional<DataVersion>, thanks

File: src/app/BufferedReadCallback.cpp
Comment (yunhanw-google) [reply src/app/BufferedReadCallback.cpp:259]: just update with optional<DataVersion>

**Comment by woody-apple** [comment]
Fast tracking, given > 3 days for review.

**Comment by woody-apple** [comment]
Fast tracking, given > 3 business days for review.

File: src/app/BufferedReadCallback.cpp
Comment (bzbarsky-apple) [inline src/app/BufferedReadCallback.cpp:229]: Followup: we should decide what we do if we have a chunked result come in and aVersion does not match mDataVersion.

File: src/app/reporting/Engine.cpp
Comment (bzbarsky-apple) [inline src/app/reporting/Engine.cpp:141]: Why?  Just because something failed to allocate memory somewhere encoding a (possibly large) data response doesn't mean we can't try to encode a status response.

File: src/app/util/ember-compatibility-functions.cpp
Comment (bzbarsky-apple) [inline src/app/util/ember-compatibility-functions.cpp:265]:
```suggestion
        ChipLogError(DataManagement, "Endpoint %" PRIx16 ", Cluster " ChipLogFormatMEI " not found in ReadClusterDataVersion!",
```

File: src/app/util/ember-compatibility-functions.cpp
Comment (bzbarsky-apple) [inline src/app/util/ember-compatibility-functions.cpp:278]:
```suggestion
        ChipLogError(DataManagement, "Endpoint %" PRIx16 ", Cluster " ChipLogFormatMEI " not found in IncreaseClusterDataVersion!",
```

File: src/app/util/ember-compatibility-functions.cpp
Comment (bzbarsky-apple) [inline src/app/util/ember-compatibility-functions.cpp:284]:
```suggestion
        ChipLogDetail(DataManagement, "Endpoint %" PRIx16 ", Cluster " ChipLogFormatMEI " update version to %" PRIx32, aEndpointId,
```

File: src/controller/tests/TestReadChunking.cpp
Comment (bzbarsky-apple) [inline src/controller/tests/TestReadChunking.cpp:81]: This part can go away with the most recent update to #14471.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/reporting/Engine.cpp
Comment (bzbarsky-apple) [reply src/app/reporting/Engine.cpp:141]: Ah, I see.  This is an attempt to chunk on the level of the whole attribute.

It seems to me that the behavior should then be as follows:

1. If error is not CHIP_ERROR_BUFFER_TOO_SMALL and is not CHIP_ERROR_NO_MEMORY, rollback and encode status.
2. Otherwise, if partial data allowed, save the encode state.
3. Otherwise roll back.  If we have already encoded some chunks, we are done; otherwise encode status.

Thoughts?

File: src/app/BufferedReadCallback.cpp
Comment (yunhanw-google) [reply src/app/BufferedReadCallback.cpp:229]: https://github.com/project-chip/connectedhomeip/issues/14530, yep

File: src/app/reporting/Engine.cpp
Comment (yunhanw-google) [reply src/app/reporting/Engine.cpp:141]: yep, agree, updated, thanks

File: src/app/util/ember-compatibility-functions.cpp
Comment (yunhanw-google) [reply src/app/util/ember-compatibility-functions.cpp:265]: Done, thanks

File: src/app/util/ember-compatibility-functions.cpp
Comment (yunhanw-google) [reply src/app/util/ember-compatibility-functions.cpp:278]: Done, thanks

File: src/app/util/ember-compatibility-functions.cpp
Comment (yunhanw-google) [reply src/app/util/ember-compatibility-functions.cpp:284]: Done, thanks

File: src/controller/tests/TestReadChunking.cpp
Comment (yunhanw-google) [reply src/controller/tests/TestReadChunking.cpp:81]: would rebase after 14471 is merged

File: src/app/reporting/Engine.cpp
Comment (yunhanw-google) [reply src/app/reporting/Engine.cpp:145]: https://github.com/project-chip/connectedhomeip/issues/14566

Comment (yunhanw-google) [reply src/app/reporting/Engine.cpp:145]: per offline discussion, file https://github.com/project-chip/connectedhomeip/issues/14566 for follow-up

**Review by yufengwangca — APPROVED** [review]

File: src/app/ReadClient.h
Comment (yufengwangca) [inline src/app/ReadClient.h:114]: nit: add spaces to align with the comments

Comment (yunhanw-google) [reply src/app/ReadClient.h]: Done, thanks

File: src/app/ReadClient.h
Comment (mrjerryjohns) [inline src/app/ReadClient.h:114]: Tag compression is an over the wire detail. By the time it gets up here to this call-site, we should _**always**_ have a data version on hand. Consequently, this should not be an `Optional`.

**Review by mrjerryjohns — CHANGES_REQUESTED** [review]

File: src/app/ReadClient.h
Comment (yunhanw-google) [reply src/app/ReadClient.h:114]: Done, thanks

File: src/app/reporting/Engine.cpp
Comment (mrjerryjohns) [inline src/app/reporting/Engine.cpp:148]: If we fail to encode the path-specific error here because we couldn't fit, will it be attempted in the next chunk attempt?

File: src/app/util/ember-compatibility-functions.cpp
Comment (mrjerryjohns) [inline src/app/util/ember-compatibility-functions.cpp:365]: Why is this call not error checked? And why do we need a notion of a temporary data version?

File: src/controller/python/chip/clusters/Attribute.py
Comment (mrjerryjohns) [inline src/controller/python/chip/clusters/Attribute.py:321]: There are a number of different 'versions' in the IM - can we be more explicit here, i.e 'dataVersion'?

File: src/controller/python/chip/clusters/Attribute.py
Comment (mrjerryjohns) [inline src/controller/python/chip/clusters/Attribute.py:557]: We're not doing much with this version, and not actually exposing this to users to see and interact with. Please file an issue to do so.

**Comment by mrjerryjohns** [comment]
Where are the checks of the data version in the ReadRequest against the version currently present on the server, and omitting data if that matches? That presumably is the biggest reason we're doing this whole version feature, which is to optimize out traffic for reads/subscribes.

File: src/app/reporting/Engine.cpp
Comment (yunhanw-google) [reply src/app/reporting/Engine.cpp:148]: yes, I think that logic has not yet been changed, it would be attempted in next chhunk

File: src/controller/python/chip/clusters/Attribute.py
Comment (yunhanw-google) [reply src/controller/python/chip/clusters/Attribute.py:557]: https://github.com/project-chip/connectedhomeip/issues/14587

File: src/app/util/ember-compatibility-functions.cpp
Comment (yunhanw-google) [reply src/app/util/ember-compatibility-functions.cpp:365]: we should do the error check, we should remove temporary data version, and use kUndefinedDataVersion, thanks

File: src/controller/python/chip/clusters/Attribute.py
Comment (yunhanw-google) [reply src/controller/python/chip/clusters/Attribute.py:321]: Done, thanks

File: src/app/reporting/Engine.cpp
Comment (yunhanw-google) [reply src/app/reporting/Engine.cpp:148]: apReadHandler->SetAttributeEncodeState(AttributeValueEncoder::AttributeEncodeState()); which would encode the last attribute data state.
in next chunk, it would load the state, AttributeValueEncoder::AttributeEncodeState encodeState = apReadHandler->GetAttributeEncodeState();

**Comment by yunhanw-google** [comment]
> Where are the checks of the data version in the ReadRequest against the version currently present on the server, and omitting data if that matches? That presumably is the biggest reason we're doing this whole version feature, which is to optimize out traffic for reads/subscribes.

DataVersionFilters has not yet been implemented in this PR, would happen in next PR, thanks

**Comment by yunhanw-google** [comment]
> > Where are the checks of the data version in the ReadRequest against the version currently present on the server, and omitting data if that matches? That presumably is the biggest reason we're doing this whole version feature, which is to optimize out traffic for reads/subscribes.
> 
> DataVersionFilters has not yet been implemented in this PR, would happen in next PR, thanks

https://github.com/project-chip/connectedhomeip/issues/14588

**Review by mrjerryjohns — APPROVED** [review]

**Pull request merged**
